### PR TITLE
ZK Session Expiry management (phase 1): Mark the node as follower

### PR DESCRIPTION
**Pull request by vmaheshw**
ZK Session expiry is not handled correctly. Calling a disconnect on zk session expiry and trying to connect again, does not really work as it both the call happens on zkClientEventThread which is already interrupted. This result in silent exit of zkClientEventThread.

Re-initializing the zkclient correctly is not the only problem. We don't have leader-> follower transition sequence in the code. As soon as session gets expired, the live instance node gets deleted and the leader is no longer participating in the election. We need to fix the transition of leader -> follower to gracefully handle zk session expiry. 

We will solve this problem in multiple commits:
1. Mark the leader as follower so that it does not do any more assignment and affect the zk state.
2. Cleanup the tasks assigned.
3. Re-initialize the same zk session.

### CODE REVIEW

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java [hunk 1]
```diff
@@ -1435,8 +1474,70 @@ public void handleDataChange(String dataPath, Object data) throws Exception {
...
+    };
+    _timer.schedule(timerTask, _sessionTimeoutMs);
+  }
+
+  // This is a temporary change to bring down the node on session expiry. We expect this to happen rarely.
+  // We are working on the transition of leader to follower on session expiry. On session expiry,
+  // the live instance (ephemeral nodes) gets deleted by the server and node automatically lose the election.
+  // By bringing down the node, the health monitoring services to restart the service is expected.
+  // This exit call will invoke the shutdownhook and will bring down the system. This is currently
+  // supported using a knob and adding shutdownhook will ensure the clean shutdown.
```
Comment (somandal) [inline datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java:1536]: Again, a bit confusing here. Not sure we need to add details about nurse bringing the node back up.  Also, not sure you really need to discuss the knob (which I assume is the config?) here. 
nit: shutdownhook -> shutdown hook

Maybe rephrase it to: Calling exit() will invoke the shutdown hook and bring down the system via a clean shutdown.

If you do want to mention the config, you can add a sentence: This behavior to bring down the system on session expiry is controlled via a config option.

If you do want to discuss the nurse job, you can add: A health monitoring service is expected to be in place to restart  the Brooklin service.

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java [hunk 2]
```diff
@@ -1435,8 +1474,70 @@ public void handleDataChange(String dataPath, Object data) throws Exception {
...
+  private void exitOnSessionExpiryOrEstablishmentError() {
+    if (_exitOnSessionExpiry) {
+      LOG.warn("Calling Exit 1");
```
Comment (somandal) [inline datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java:1539]: nit: for consistency prepend the comment with "ZkStateChangeListener::"?

File: datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java [hunk 3]
```diff
@@ -715,9 +725,76 @@ ZkClient createZkClient() {
...
+    tasks.add(task);
+    updateInstanceAssignment(adapter, adapter.getInstanceName(), tasks);
+
+    LOG.info("Acquire from instance1 should succeed");
+    Assert.assertTrue(expectException(() -> task.acquire(timeout), false));
+
+    simulateSessionExpiration(adapter);
+
+    Thread.sleep(5000);
+    Assert.assertTrue(adapter.getZkStateChangeListener().sessionExpired);
```
Comment (somandal) [inline datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java:779]: It will be helpful to validate that exitOnSessionExpiryOrEstablishmentError() is indeed called on session expiry

File: datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java [hunk 4]
```diff
@@ -715,9 +725,76 @@ ZkClient createZkClient() {
...
     public ZkClient getZkClient() {
       return _zkClient;
     }
+
+    public ZkClientMockStateChangeListener getZkStateChangeListener() {
+      return _zkClientMockStateChangeListener;
+    }
+  }
+
+  @Test
```
Comment (somandal) [inline datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java:755]: Is is possible to simulate the "Disconnected" state? If so, can we add a test for that, and validate that scheduleExpiryTimerAfterSessionTimeout() is indeed called too? Then validate two cases, a) Let it expire and validate that exitOnSessionExpiryOrEstablishmentError() is called, b) Don't let it expire, and validate that exitOnSessionExpiryOrEstablishmentError()  is not called.

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java [hunk 5]
```diff
@@ -148,18 +157,35 @@
...
+   * @param sessionTimeoutMs Session timeout to use for the connection with the ZooKeeper server
+   * @param connectionTimeoutMs Connection timeout to use for the connection with the ZooKeeper server
+   * @param listener ZKAdapterListener implementation to receive callbacks based on various znode changes
+   * @param exitOnSessionExpiry boolean to decide whether to exit on session expiry or not.
+   */
+  public ZkAdapter(String zkServers, String cluster, String defaultTransportProviderName, int sessionTimeoutMs,
+      int connectionTimeoutMs, ZkAdapterListener listener, boolean exitOnSessionExpiry) {
+    this(zkServers, cluster, defaultTransportProviderName, sessionTimeoutMs, connectionTimeoutMs, -1,
+        listener, exitOnSessionExpiry);
+  }
```
[the remark this answers is not in the document]
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java:188]: My intention is to not touch all the existing callers of current constructors. Instead add a new one and only modify one place. This will also makes the sequence of constructors cleaner.

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java [hunk 6]
```diff
@@ -222,14 +248,22 @@ ZkClient createZkClient() {
...
+    }
+    return _stateChangeListener;
+  }
+
   /**
    * Connect the adapter so that it can connect and bridge events between ZooKeeper changes and
    * the actions that need to be taken with them, which are implemented in the Coordinator class
    */
   public void connect() {
     disconnect(); // Guard against leaking an existing zookeeper session
```
[the remark this answers is not in the document]
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java:264]: I will do it in next pass. Removing it will make the situation worse, if this PR has any bugs.

File: datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java [hunk 7]
```diff
@@ -259,7 +258,8 @@ public Coordinator(CachedDatastreamReader datastreamCache, CoordinatorConfig con
   @VisibleForTesting
```
[the remark this answers is not in the document]
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java:258]: :) Even I don't see this getting used using intellij. But I remember adding it. This is used in TestCoordinatorWithSpyZkAdapter in TestCoordinator.java. Probably some intellij bug.

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java [hunk 8]
```diff
@@ -1435,8 +1474,70 @@ public void handleDataChange(String dataPath, Object data) throws Exception {
...
+   */
+  @VisibleForTesting
+  class ZkStateChangeListener implements IZkStateListener {
+    @Override
+    public void handleStateChanged(Watcher.Event.KeeperState state) {
+      LOG.info("ZkStateChangeListener:: handleStateChanged {}", state.toString());
+      switch (state) {
+        case Expired:
+          _timer.cancel();
+          onBecomeFollower();
```
[the remark this answers is not in the document]
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java:1493]: Yes, Exit will call the shutdown hook, which will call coordinator.stop(). This will be a safe practice.

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java [hunk 9]
```diff
@@ -1435,8 +1474,70 @@ public void handleDataChange(String dataPath, Object data) throws Exception {
...
+          exitOnSessionExpiryOrEstablishmentError();
+          return;
+        case Disconnected:
+          // if the session has expired it means that all the registration's ephemeral nodes are gone.
```
[the remark this answers is not in the document]
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java:1497]: I wanted to keep this generic. I have put the explanation in the exit call. I have cleaned this particular comment to be more specific to general zookeeper behavior.

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java [hunk 10]
```diff
@@ -1435,8 +1474,70 @@ public void handleDataChange(String dataPath, Object data) throws Exception {
...
+    }
+
+    @Override
+    public void handleSessionEstablishmentError(final Throwable error) {
+      LOG.error("ZkStateChangeListener:: Failed to establish session.", error);
+      exitOnSessionExpiryOrEstablishmentError();
+    }
+  }
+
+  private void scheduleExpiryTimerAfterSessionTimeout() {
```
[the remark this answers is not in the document]
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java:1522]: I wanted to highlight that the timer is set of the value session timeout after which we will call the session expired routine.

File: datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java (hunk 4 above)
Comment (vmaheshw) [reply datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java:755]: No, I tried a lot and could not find a way. I couldn't find it in samza repo as well.

File: datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java [hunk 11]
```diff
@@ -716,9 +726,77 @@ ZkClient createZkClient() {
...
+    ZkClientInterceptingAdapter adapter = createInterceptingZkAdapter(testCluster, 5000);
+    adapter.connect();
+
+    DatastreamTaskImpl task = new DatastreamTaskImpl();
+    task.setId("3");
+    task.setConnectorType(connectorType);
+    task.setZkAdapter(adapter);
+
+    List<DatastreamTask> tasks = new ArrayList<>();
+    tasks.add(task);
```
Comment (ahmedahamid) [inline datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java:771]: `Collections.singletonList()`?

File: datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java [hunk 12]
```diff
@@ -716,9 +726,77 @@ ZkClient createZkClient() {
...
+    updateInstanceAssignment(adapter, adapter.getInstanceName(), tasks);
+
+    LOG.info("Acquire from instance1 should succeed");
+    Assert.assertTrue(expectException(() -> task.acquire(timeout), false));
+
+    simulateSessionExpiration(adapter);
+
+    Thread.sleep(5000);
```
Comment (ahmedahamid) [inline datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java:779]: Please excuse my ignorance: why do we have to sleep for that long? Is it because this is the value we specified for `sessionTimeoutMs`? If that's the reason, should we decrease that value?

File: datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java [hunk 13]
```diff
@@ -716,9 +726,77 @@ ZkClient createZkClient() {
...
+    } catch (NoSuchFieldException | IllegalAccessException e) {
+      Assert.fail(e.toString());
+    }
+
+    ZooKeeper zookeeper = zkConnection.getZookeeper();
+    long sessionId = zookeeper.getSessionId();
+
+    LOG.info("Closing/expiring session:" + sessionId);
+    ZooKeeperServer zkServer = _embeddedZookeeper.getZooKeeperServer();
+    zkServer.closeSession(sessionId);
```
Comment (ahmedahamid) [inline datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java:799]: Unless you foresee a good reason to keep it the way it is, I'd suggest exposing a `closeSession()` on `EmbeddedZookeeper` to limit the exposed internals.

File: datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java (hunk 12 above)
Comment (vmaheshw) [reply datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java:779]: I gave a decent value so that it does not fail on pcl machines which are slower and then this test will become flaky.

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java:240]: If you call disconnect without connect, it can lead to leaderless scenario till we fix everything. [previous analysis](https://docs.google.com/document/d/14rhn6U6AOqZ_swkQsuUKzH7DCS7Ab3aSvWcuXhESBEI/edit#heading=h.syjnl87mxqob) where we had zk sessions leak when disconnect was not called before connect.

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java (hunk 1 above)
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java]: ignoring, since not doing this now.

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java [hunk 14]
```diff
@@ -222,14 +229,22 @@ ZkClient createZkClient() {
...
+
   /**
    * Connect the adapter so that it can connect and bridge events between ZooKeeper changes and
    * the actions that need to be taken with them, which are implemented in the Coordinator class
    */
   public void connect() {
     disconnect(); // Guard against leaking an existing zookeeper session
     _zkclient = createZkClient();
-
+    _zkclient.subscribeStateChanges(getOrCreateStateChangeListener());
```
[the remark this answers is not in the document]
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java:242]: Tried to handle as many as I can.

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java [hunk 15]
```diff
@@ -334,6 +349,12 @@ private void joinLeaderElection() {
 
     // if this instance is first in line to become leader. Check if it is already a leader.
     if (index == 0) {
+      // The node at index 0 is a leader and does not have to listen on a previous node anymore.
+      if (_currentSubscription != null) {
+        _zkclient.unsubscribeDataChanges(_currentSubscription, _leaderElectionListener);
```
[the remark this answers is not in the document]
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java]: Fixed.

File: datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java [hunk 16]
```diff
@@ -716,9 +726,77 @@ ZkClient createZkClient() {
...
+    try {
+      Field privateField = ZkClient.class.getSuperclass().getDeclaredField("_connection");
+      privateField.setAccessible(true);
+      zkConnection = (ZkConnection) privateField.get(adapter.getZkClient());
+    } catch (NoSuchFieldException | IllegalAccessException e) {
+      Assert.fail(e.toString());
+    }
+
+    ZooKeeper zookeeper = zkConnection.getZookeeper();
+    long sessionId = zookeeper.getSessionId();
```
[the remark this answers is not in the document]
Comment (vmaheshw) [reply datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java]: Done.

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java (hunk 2 above)
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java]: not required now

File: datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java (hunk 3 above)
Comment (vmaheshw) [reply datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java:775]: Not required now.

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java:242]: Done.

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java [hunk 17]
```diff
@@ -1265,13 +1286,13 @@ public static String parseHostnameFromZkInstance(String instance) {
...
    */
-  public class ZkBackedDMSDatastreamList implements IZkChildListener, IZkDataListener {
+  private class ZkBackedDMSDatastreamList implements IZkChildListener, IZkDataListener {
     private final String _path;
 
     /**
      * Sets up a watch on the {@code /{cluster}/dms} tree, so it can be notified of future changes.
      */
-    public ZkBackedDMSDatastreamList() {
+    private ZkBackedDMSDatastreamList() {
```
[the remark this answers is not in the document]
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java:1316]: Done

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java [hunk 18]
```diff
@@ -1331,8 +1353,8 @@ public void handleDataDeleted(String dataPath) throws Exception {
    * the new leader is responsible for cleaning up the instance node for the previous leader. This is
    * done in the constructor ZkBackedLiveInstanceListProvider().
    */
-  public class ZkBackedLiveInstanceListProvider implements IZkChildListener {
-    private List<String> _liveInstances = new ArrayList<>();
+  private class ZkBackedLiveInstanceListProvider implements IZkChildListener {
+    private List<String> _liveInstances;
     private final String _path;
```
[the remark this answers is not in the document]
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java:1379]: Done.

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java [hunk 19]
```diff
@@ -1371,12 +1393,12 @@ public void close() {
       _zkclient.unsubscribeChildChanges(_path, this);
     }
 
-    public List<String> getLiveInstances() {
+    private List<String> getLiveInstances() {
```
[the remark this answers is not in the document]
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java:1417]: Done

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java [hunk 20]
```diff
@@ -1453,15 +1476,15 @@ public void handleDataChange(String dataPath, Object data) throws Exception {
...
       // do nothing
     }
   }
 
   /**
    * ZkTargetAssignmentProvider detect if there is a partition movement being intiated from restful endpoint
    */
-  public class ZkTargetAssignmentProvider implements IZkDataListener {
+  private class ZkTargetAssignmentProvider implements IZkDataListener {
     Set<String> _listenedConnectors = new HashSet<>();
```
[the remark this answers is not in the document]
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java:1509]: Done

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java [hunk 21]
```diff
@@ -1508,8 +1531,61 @@ public void handleDataChange(String dataPath, Object data) throws Exception {
...
+  /**
+   * Listener for ZooKeeper state changes.
+   */
+  @VisibleForTesting
+  class ZkStateChangeListener implements IZkStateListener {
+    @Override
+    public void handleStateChanged(Watcher.Event.KeeperState state) {
+      LOG.info("ZkStateChangeListener::handleStateChanged {}", state.toString());
+      switch (state) {
+        case Expired:
```
[the remark this answers is not in the document]
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java:1570]: We are cancelling the timer set during disconnect in expiry path. 

If we get to know if session expiry from disconnect  + timer expiry before notification from state change: yes
If we get to know if session expiry from state change before disconnect + timer expiry: no


. But to safe guard

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java [hunk 22]
```diff
@@ -1508,8 +1531,61 @@ public void handleDataChange(String dataPath, Object data) throws Exception {
...
+          // Wait for session timeout after disconnect to consider that the session has expired.
+          LOG.warn("ZkStateChangeListener::Got {} event. Scheduling a system stop.", state.toString());
+          scheduleExpiryTimerAfterSessionTimeout();
+          return;
+        case SyncConnected:
+          LOG.info("ZkStateChangeListener::Connected. Canceling timer.");
+          _timer.cancel();
+          return;
+        default:
+          // Ignoring AuthFailed for now.
```
[the remark this answers is not in the document]
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java:1584]: The other states are logged by ClientCxn as well. It will be duplicate log.

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java [hunk 23]
```diff
@@ -1508,8 +1531,61 @@ public void handleDataChange(String dataPath, Object data) throws Exception {
...
+   * Listener for ZooKeeper state changes.
+   */
+  @VisibleForTesting
+  class ZkStateChangeListener implements IZkStateListener {
+    @Override
+    public void handleStateChanged(Watcher.Event.KeeperState state) {
+      LOG.info("ZkStateChangeListener::handleStateChanged {}", state.toString());
+      switch (state) {
+        case Expired:
+          _timer.cancel();
```
[the remark this answers is not in the document]
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java:1571]: Thats a good catch. Fixed it.

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (ahmedahamid) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java:235]: Got it. How about resetting the current thread's interrupted flag after calling `disconnect()` so the subsequent `createZkClient()` can proceed?

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java [hunk 24]
```diff
@@ -1508,8 +1547,89 @@ public void handleDataChange(String dataPath, Object data) throws Exception {
...
+    public void handleDataDeleted(String dataPath) {
       // do nothing
     }
   }
+
+  /**
+   * Listener for ZooKeeper state changes.
+   */
+  private class ZkStateChangeListener implements IZkStateListener {
+    private Timer _timer;
```
Comment (ahmedahamid) [inline datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java:1559]: I think the code would be a lot simpler if we used a `ScheduledExecutorService` instead of this `Timer`. 

Here's what I'm thinking:
- We declare two fields:
    ```java
        private final ScheduledExecutorService _scheduledExecutorService;
        private Future<?> _zkSessionExpiryFuture = CompletableFuture.completedFuture(null);
    ```
- We change `cancelTimer()` to do:
   ```java
   _zkSessionExpiryFuture.cancel(false);
   ```
- We change `scheduleExpiryTimerAfterSessionTimeout()` to do:
    ```java
     _zkSessionExpiryFuture =
          _scheduledExecutorService.schedule(ZkAdapter.this::onSessionExpired, _sessionTimeoutMs,
              TimeUnit.MILLISECONDS);
    ```

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java:235]: Adding this change as part of this PR can open another set of issues. The behavior will not change much after the current PR. I will consider this when we are ready to enable reconnect and sure that there are no other loose ends.

[the remark this answers is not in the document]
Comment (ahmedahamid) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java:235]: OK. So, just to confirm my understanding, this PR is not meant to handle the following scenario:
1. ZK session expires. Out code observes a `Disconnected` state and schedules an `onSessionExpired()` call
2. `onSessionExpired()` gets invoked after the timeout elapses and `onBecomeFollower()` is called
3. ZK client reestablishes connectivity and its connection state transitions to `Expired`
4. ZK client's `fireAllEvents()` is invoked, calls `joinLeaderElection()`
5. The leader election code fails to find the current instance name among the live instances, calls `connect()` which calls `disconnect()` followed by `createZkClient()` — an exception is thrown and the host gets in an undefined state.

[the remark this answers is not in the document]
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java:235]: The sequence of the scenario is right, but we have more clarity on point 5.

1. Leader node will not run as leader node anymore (So, we will not see multiple leaders problem)
2. We will have metrics in next PR that we will use as alert to restart the node (which is not happening today)
3. The zookeeper states will be fine.

Next step will be:
1. Add metrics and alert.
2. Task cleanup support within fixed time.
3. Add support to handle low-level kafka.
4. Then enable the reconnection.

[the remark this answers is not in the document]
Comment (somandal) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java:235]: @vmaheshw Personally, it'll be much easier to review any changes related to ZK session expiry once I've seen your design and implementation plan. It can avoid unnecessary confusion like this. It'd be nice if you can prioritize working on that before sending out further PRs. It will also help with prioritizing the next steps.

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java (hunk 24 above)
Comment (somandal) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java:1559]: +1, I like this suggestion more than using the `_timer`

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java [hunk 25]
```diff
@@ -1508,8 +1547,89 @@ public void handleDataChange(String dataPath, Object data) throws Exception {
...
+      switch (state) {
+        case Expired:
+          cancelTimer();
+          onSessionExpired();
+          return;
+        case Disconnected:
+          // Wait for session timeout after disconnect to consider that the session has expired.
+          LOG.warn("ZkStateChangeListener::Got {} event. Scheduling a system stop.", state);
+          scheduleExpiryTimerAfterSessionTimeout();
+          resetSession.set(true);
```
[the remark this answers is not in the document]
Comment (somandal) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java:1577]: +1 I had the same question when going through the code. I like Ahmed's suggestion here.

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java (hunk 24 above)
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java]: Done.

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java (hunk 25 above)
[the remark this answers is not in the document]
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java]: Done.

File: datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java [hunk 26]
```diff
@@ -1508,8 +1547,89 @@ public void handleDataChange(String dataPath, Object data) throws Exception {
...
+      if (_timer == null) {
+        _timer = new Timer(true);
+      }
+      TimerTask timerTask = new TimerTask() {
+        public void run() {
+          if (resetSession.get()) {
+            onSessionExpired();
+          }
+        }
+      };
```
[the remark this answers is not in the document]
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java]: Done.
